ASTC texture support

bmdhacks opened the pull request:
ASTC is a modern texture compression format that is ubiquitously supported in almost every mobile processor.  It provides better quality, better compression, and more options than all other formats and has emerged as the industry standard for compressed textures in mobile devices.

This patch adds and extends ASTC support in OSG by allowing ASTC blocks to be wrapped in a DDS container.  This is nonstandard, as ASTC is more of an OpenGL technology and DDS is a DirectX container, so I had to take some liberties.  First, I straight up made up my own DXGI format numbers for all the ASTC block sizes.  I think Apple's Metal API might use these same numbers in their DDS/ASTC support but I'm not sure.  Secondly, I wrap the ASTC image data with DDS using OpenGL addressing of bottom-left first.  This means that all textures will be displayed inverted in OpenMW unless  you encode them flipped to begin with.  ASTC block data is to complex to flip at runtime like we do with DXT1 and DXT5 so you just have to encode the texture inverted.

Because this is nonstandard, I hacked up some random tool I found on github called kram to do the encoding for me.  You can find this here:
https://github.com/bmdhacks/kram
It would be nice to have better tooling for creation and conversion of DDS textures in OpenMW but this is good enough for me to get something out the door.

Capostrophic commented:
Yeah, I don't like that wrapping it into DDS. We and you shouldn't want to effectively maintain not just an extension of DDS *but also* its relevant tooling. Maybe add it as a separate astc format? Like whatever is done [here](https://github.com/ARM-software/astc-encoder/tree/main)? That way existing tooling for ASTC should be possible to be used.

bmdhacks committed:
KTX ASTC Support

I removed some vestiges of ASTC support in DDS, and fixed a few math bugs regarding pixel sizes and compressed size for astc.
Also, we don't try to invert ASTC KTX textures becaue they're fed to opengl correctly as long as the origin is set accurately.

Capostrophic commented on src/osgPlugins/ktx/ReaderWriterKTX.cpp:
```diff
@@ -210,6 +210,9 @@ osgDB::ReaderWriter::ReadResult ReaderWriterKTX::readKTXStream(std::istream& fin
         header.glType, totalImageData, osg::Image::USE_NEW_DELETE);
 
+    // KTX textures are stored in their original coordinate system (TOP_LEFT)
+    image->setOrigin(osg::Image::TOP_LEFT);
```
Apparently not. It's primarily determined by KTXorientation key value pair for both KTX and KTX2, but the recommended "default" is top-left.  

This still should not be set, though, as OSG doesn't support KTXorientation and handles everything using OpenGL origin.

bmdhacks replied:
The good news is that in OSG, image->setOrigin is a noop function.

bmdhacks committed:
Look at the ktxOrientation metadata field for ktx textures in order to ensure compatible textures.

bmdhacks committed:
make badly aligned KTX textures just a warning

bmdhacks commented:
Ok, I'm parsing the ktx orientation metadata and warning if it is incorrect.  I've also modified all my conversion utils (and have a PR to astcenc) to set this metadata field.  I think this is committable now.

AnyOldName3 commented:
I can absolutely see the desire for a warning for a missing/incompatible orientation, but I have mixed feelings about having it in this PR. I guess part of this is because it should really be OpenMW's job to decide it doesn't support certain orientations (or is unlikely to support things with unspecified orientation) rather than OSG's, but as the orientation enum doesn't have an *unknown* value, we can't really pass this information to the engine, so it's kind of a moot point. I've sat on the problem for a day without thinking of anything better than what this PR does, so maybe it's already as good as it's going to get.

bmdhacks commented:
Yeah, the way I see it, we have three options.  I'm not passionate about any of them:

1. Allow all KTX textures and let the user figure out that it's OpenGL oriented, and thus probably needs to be inverted.
2. Warn if the KTX texture isn't OpenGL oriented with proper metadata, which incentivizes the user to use one of our encoding tools (although hopefully astcenc will accept my PR) but if they really want to use Photoshop and just flip the image that'll at least work for testing.
3. Fail to display any texture that doesn't have the correct metadata.  As far as I can tell, this will mean they must use our tools because most ktx encoders don't set the ktxOrientation flag, and nobody seems to read it either (despite it being in the spec).

I went with #2 in this PR, but I'm happy to change it if y'all would prefer.  All the portmaster tools will add the metadata to our textures so at least we'll be compliant with any option.

bmdhacks commented:
Ok i can add it there but that seems like the wrong place in the separation
of concerns for me. In my view of the design, OSG should handle textues and
openmw should just view them abstractly.

On Mon, Sep 8, 2025, 9:45 AM Alexei Kotov ***@***.***> wrote:

> *Capostrophic* left a comment (OpenMW/osg#31)
> <https://github.com/OpenMW/osg/pull/31#issuecomment-3267128935>
>
> We already have some format-specific logic e.g. for TGA. The image manager
> will know if we used the KTX plugin to parse the image, and then we can try
> and do something with the result image.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenMW/osg/pull/31#issuecomment-3267128935>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHW2K35HTN3A4IFE4MTP7D3RWXDDAVCNFSM6AAAAACEV5TXIGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENRXGEZDQOJTGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Capostrophic commented:
It's impossible for OSG to properly adjust the image to match the expected orientation in every single case. It's not unreasonable for it to provide the raw data in the osg::Image with no modifications and **no warnings**. The problem is with providing the orientation for the application to be able to properly interpret the osg::Image. Right now extending osg::Image to provide more orientations comes with complications. Whether or not the extra orientations are added, it would be the responsibility of the application to warn the user it cannot properly handle the supplied orientation.

We have full info about the state of KTX images by the point they are converted because OSG is not maintained, so we would be able to know for sure top left origin osg::Image is not the orientation we expect.

AnyOldName3 commented:
The TGA logic isn't something we *want* to have, it's something we're forced to have because Morrowind handles TGA files weirdly and some mods don't work properly unless we inject the same weirdness.

Maybe something we could do is put the key/value data from the file into the `osg::Image`'s userdata. In principle, application-specific userdata in a KTX file could do other application-specific things and this seems to me like something that's potentially sensible for other reasons.

Capostrophic commented:
Using the user data seems sane. Although we'd still need to distinguish supported non-KTX images with no user data from unsupported KTX images with no user data.

bmdhacks committed:
store ktx metdata in the images user data structure and remove the warning about orientation.

bmdhacks commented:
Ok, I'm storing ktx metadata in the image user store now with the "KTX:" prefix.  No warnings are emitted if the texture doesn't have the correct orientation, I'll do that in openmw.

bmdhacks committed:
remove a noop
